arm9/lib/libnns/src/NNS_FND_expheap.c

```diff
@@ -0,0 +1,385 @@
+    NNSiFndExpHeapMBlockHead* pMBlkHd = NULL;
+    NNSiFndExpHeapMBlockHead* pMBlkHdFound = NULL;
+    u32 foundSize = 0xFFFFFFFF;
+    void* foundMBlock = NULL;
+
+    for (pMBlkHd = pExpHeapHd->mbFreeList.tail; pMBlkHd; pMBlkHd = pMBlkHd->pMBHeadPrev)
+    {
+        void *const mblock = GetMemPtrForMBlock(pMBlkHd);
+        void *const mBlockEnd = AddU32ToPtr(mblock, pMBlkHd->blockSize);
+        void *const reqMBlock = NNSi_FndRoundDownPtr(SubU32ToPtr(mBlockEnd, size), alignment);
+        if (ComparePtr(reqMBlock, mblock) >= 0 && foundSize > pMBlkHd->blockSize)
+        {
+            pMBlkHdFound = pMBlkHd;
+            foundSize = pMBlkHd->blockSize;
+            foundMBlock = reqMBlock;
+            if (bAllocFirst || foundSize == size)
+                break;
+        }
+    }
+
+    if (!pMBlkHdFound)
+        return NULL;
+
+    return AllocUsedBlockFromFreeBlock(pExpHeapHd, pMBlkHdFound, foundMBlock, size, 1);
+}
+
+BOOL RecycleRegion(NNSiFndExpHeapHead* pEHHead, const NNSiMemRegion* pRegion)
+{
+    NNSiFndExpHeapMBlockHead* pBlkPtrFree = NULL;
+    NNSiMemRegion freeRgn = *pRegion;
+    NNSiFndExpHeapMBlockHead* pBlk;
+    for (pBlk = pEHHead->mbFreeList.head; pBlk; pBlk = pBlk->pMBHeadNext)
+    {
+        if (pBlk < (NNSiFndExpHeapMBlockHead*)pRegion->start)
+        {
+            pBlkPtrFree = pBlk;
+            continue;
+        }
+        if (pBlk == pRegion->end)
+        {
+            freeRgn.end = GetMBlockEndAddr(pBlk);
+            (void)RemoveMBlock(&pEHHead->mbFreeList, pBlk);
+        }
+        break;
+    }
+    if (pBlkPtrFree && GetMBlockEndAddr(pBlkPtrFree) == pRegion->start)
+    {
+        freeRgn.start = pBlkPtrFree;
+        pBlkPtrFree = RemoveMBlock(&pEHHead->mbFreeList, pBlkPtrFree);
+    }
+    if (GetOffsetFromPtr(freeRgn.start, freeRgn.end) < sizeof(NNSiFndExpHeapMBlockHead))
+        return FALSE;
+    InsertMBlock(&pEHHead->mbFreeList, InitFreeMBlock(&freeRgn), pBlkPtrFree);
+    return TRUE;
+}
+
+NNSFndHeapHandle NNS_FndCreateExpHeapEx(void *startAddress, u32 size, u16 optFlag)
+{
+    void* endAddress = NNSi_FndRoundDownPtr(AddU32ToPtr(startAddress, size), 4);
+    startAddress = NNSi_FndRoundUpPtr(startAddress, 4);
+    if (NNSiGetUIntPtr(startAddress) > NNSiGetUIntPtr(endAddress) || GetOffsetFromPtr(startAddress, endAddress) < sizeof(NNSiFndHeapHead) + sizeof(NNSiFndExpHeapHead) + sizeof(NNSiFndExpHeapMBlockHead) + 4)
+        return NULL;
+    return InitExpHeap(startAddress, endAddress, optFlag);
+}
+
+void NNS_FndDestroyExpHeap(NNSFndHeapHandle handle)
+{
+    NNSi_FndFinalizeHeap(handle);
+}
+
+void* NNS_FndAllocFromExpHeapEx(NNSFndHeapHandle handle, u32 size, int alignment)
+{
+    if (size == 0)
+        size = 1;
+    size = NNSi_FndRoundUp(size, 4);
+    if (alignment >= 0)
+        return AllocFromHead(handle, size, alignment);
+    else
+        return AllocFromTail(handle, size, -alignment);
+}
+
+/*
+u32 NNS_FndResizeForMBlockExpHeap(NNSFndHeapHandle heap, void *memBlock, u32 size)

```

Comment:
commented out? does this not match?

Reply:
yeah, i couldn't match it, that's part of the reason why i'm making this a stopping point. there's still the rest of expheap and all of frameheap left in NNS fnd